fix: Make selfHeal=true work when previous sync fails (#18442)

mbaldessari opened the pull request:
Currently even when selfHeal is set to true, whenever
attemptPhase.Successful() returns false, the application will be stuck
forever and sync will never be attempted again. This contradicts
https://argo-cd.readthedocs.io/en/stable/user-guide/auto_sync/#automated-sync-semantics
where we state that:
```
Automated sync will only attempt one synchronization per unique
combination of commit SHA1 and application parameters. If the most
recent successful sync in the history was already performed against the
same commit-SHA and parameters, a second sync will not be attempted,
unless selfHeal flag is set to true.
```

Let's fix that by only avoiding another sync when selfHeal is not set to
true.

Assuming alreadyAttempted is true here are the before and after
scenarios. Possible outcomes labels in (A,B,C):

*Before*

1. selfHeal is false and attemptPhase.Successful() is false ->
   (A) return SyncError "Skipping auto-sync: failed previous sync attempt to %s"
2. selfHeal is false and attemptPhase.Successful() is true ->
   (B) return nil, 0 "Skipping auto-sync: most recent sync already to %s"
3. selfHeal is true and attemptPhase.Successful() is false ->
   (A) return SyncError "Skipping auto-sync: failed previous sync attempt to %s"
4. selfHeal is true and attemptPhase.Successful() is true ->
   (C) return either nil, 0 if shouldSelfHeal is false or set
       op.Sync.Resources if shouldSelfHeal is true (C)

*After*

1. selfHeal is false and attemptPhase.Successful() is false ->
   (A) return SyncError "Skipping auto-sync: failed previous sync attempt to %s"
2. selfHeal is false and attemptPhase.Successful() is true ->
   (B) return nil, 0 "Skipping auto-sync: most recent sync already to %s"
3. selfHeal is true and attemptPhase.Successful() is false ->
   (C) return either nil, 0 if shouldSelfHeal is false or set
       op.Sync.Resources if shouldSelfHeal is true (C)
4. selfHeal is true and attemptPhase.Successful() is true ->
   (C) return either nil, 0 if shouldSelfHeal is false or set
       op.Sync.Resources if shouldSelfHeal is true (C)

Closes: #18442

<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

agaudreault commented on controller/appcontroller.go:
```diff
@@ -1874,7 +1874,8 @@ func (ctrl *ApplicationController) autoSync(app *appv1.Application, syncStatus *
 	// and parameter overrides are different from our most recent sync operation.
-	if alreadyAttempted && (!selfHeal || !attemptPhase.Successful()) {
+	// Unless selfHeal if set to true in which case we will retry
+	if alreadyAttempted && !selfHeal {
```
If think the if logic could be made more readable. 
```
if alreadyAttempted {
  if attemptPhase.Successful() {
     // log "Skipping auto-sync: most recent sync already to %s"
     return
  }
  if !selfheal {
    // log "Skipping auto-sync: failed previous sync attempt to %s"
    return
  }
  if !shouldSelfheal {
    //log "Skipping auto-sync: already attempted sync to %s with timeout %v (retrying in %v)"
    return
  }
  // do self heal
}
```

beekhof replied:
Maybe, but perhaps the immediate focus should be fixing the rather obvious bug.

agaudreault reviewed (changes requested):
Thanks for the contribution 👍  I think this is an issue that has been there for a long time. The performance will most likely depend on `shouldSelfHeal` function having a good backoff mechanism. I see that it is not currently tested. Since your code change will most likely cause more calls to this function, do you mind adding test to it as well to make sure it is behaving properly?

andrii-korotkov-verkada reviewed (commented):
I'm not sure it's worth doing failed syncs over and over again even if self-heal is true. This would have performance implications, but I don't see how it'd help.

mbaldessari committed:
fix: allow selfHeal to retry failed sync attempts (#18442)

Currently even when selfHeal is set to true, whenever the previous sync
attempt failed, the application will be stuck forever and sync will
never be attempted again. This contradicts the documented automated sync
semantics where we state that:

  Automated sync will only attempt one synchronization per unique
  combination of commit SHA1 and application parameters. If the most
  recent successful sync in the history was already performed against
  the same commit-SHA and parameters, a second sync will not be
  attempted, unless selfHeal flag is set to true.

Fix this by only skipping a retry when selfHeal is not enabled. When
selfHeal is true, a failed previous sync now falls through to the
self-heal logic (selfHealRemainingBackoff / selfHealTimeout or
configured backoff) instead of returning an error and giving up
permanently.

Assuming alreadyAttempted is true, here are the before and after
scenarios:

Before:
  1. selfHeal=false, lastAttemptedPhase failed    -> (A) SyncError
  2. selfHeal=false, lastAttemptedPhase succeeded -> (B) skip
  3. selfHeal=true,  lastAttemptedPhase failed    -> (A) SyncError (bug)
  4. selfHeal=true,  lastAttemptedPhase succeeded -> (C) self-heal

After:
  1. selfHeal=false, lastAttemptedPhase failed    -> (A) SyncError
  2. selfHeal=false, lastAttemptedPhase succeeded -> (B) skip
  3. selfHeal=true,  lastAttemptedPhase failed    -> (C) self-heal
  4. selfHeal=true,  lastAttemptedPhase succeeded -> (C) self-heal

Closes: #18442

Signed-off-by: Michele Baldessari <michele@acksyn.org>

mbaldessari commented:
@agaudreault Apologies for dropping the ball here, I was moved to another project and only now resumed things with argo. So I updated and retested everything, the change seems to be even simpler now (thanks for your semplification in abbdfa26f, btw). I see that now the backoff has already been implemented in dc27102cf, so I think this should be sufficient for the issue?

mbaldessari committed:
Adapt tests for the fixed selfHeal behaviour

Signed-off-by: Michele Baldessari <michele@acksyn.org>

agaudreault commented:
Closing due to https://github.com/argoproj/argo-cd/issues/18442#issuecomment-4389724265.
The current behaviour is the intended one.
